source/extensions/certificate_providers/certificate_provider.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+
+#include "openssl/ssl.h"
+#include "openssl/x509v3.h"
+
+
+namespace Envoy {
+namespace Extensions {
+namespace CertificateProviders {
+
+struct Certpair {
+    X509 cert_;
+    EVP_PKEY key_;
+};
+
+struct Capabilites {
+    /* whether or not a provider provides a ca cert directly */
+    bool provide_ca_cert = false;
+
+    /* Whether or not a provider provides identity certs directly */
+    bool provide_identity_certs = false;
+
+    /* whether or not a provider supports generating identity certs during handshake,
+     * which requires the capability provide_ca_cert to sign the certificates
+     */
+    bool generate_identity_certs = false;
+};
+
+class CertificateProvider {
+public:
+    virtual ~CertificateProvider() = default;
+
+    virtual Capabilites capabilities() const PURE;
+
+    /**
+     * @return CA certificate from provider
+     */
+    virtual const std::string& getCACertificate() const PURE;
+
+    /*
+     * Get TLS identity certificates directly from provider
+     */
+    virtual std::list<Certpair> getTlsCertificates() PURE;
+
+    /*
+     * Generate TLS identity certificate dynamicly during TLS handshake
+     */
+    virtual Certpair* generateTlsCertificate(absl::string_view server_name) PURE;
```

Comment:
Why do you need `Certpair` here? The method name is `generateTlsCertificate` and you can generate TLS certificate without generating a key. Also need to clarify ownership of the returned pointer.

Reply:
This provider needs to cover two usecases:
1. tls_certificate_provider_instance in [CommonTlsContext](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/tls.proto#extensions-transport-sockets-tls-v3-commontlscontext) to provide identity certs used in TLS handshake, just like [extensions.transport_sockets.tls.v3.TlsCertificate](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#envoy-v3-api-msg-extensions-transport-sockets-tls-v3-tlscertificate), that's why generateTlsCertificate has to return both certificate and private key
2. ca_certificate_provider_instance in [CertificateValidationContext](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-certificatevalidationcontext) to provide ca certs used for validation, corresponds to function getCACertificate